// @sapphire/framework/dist/lib/plugins/PluginManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginManager = void 0;
const Enums_1 = require("../types/Enums");
const symbols_1 = require("./symbols");
class PluginManager {
    constructor() {
        this.registry = new Set();
    }
    registerHook(hook, type, name) {
        if (typeof hook !== 'function')
            throw new TypeError(`The provided hook ${name ? `(${name}) ` : ''}is not a function`);
        this.registry.add({ hook, type, name });
        return this;
    }
    registerPreGenericsInitializationHook(hook, name) {
        return this.registerHook(hook, Enums_1.PluginHook.PreGenericsInitialization, name);
    }
    registerPreInitializationHook(hook, name) {
        return this.registerHook(hook, Enums_1.PluginHook.PreInitialization, name);
    }
    registerPostInitializationHook(hook, name) {
        return this.registerHook(hook, Enums_1.PluginHook.PostInitialization, name);
    }
    registerPreLoginHook(hook, name) {
        return this.registerHook(hook, Enums_1.PluginHook.PreLogin, name);
    }
    registerPostLoginHook(hook, name) {
        return this.registerHook(hook, Enums_1.PluginHook.PostLogin, name);
    }
    use(plugin) {
        const possibleSymbolHooks = [
            [symbols_1.preGenericsInitialization, Enums_1.PluginHook.PreGenericsInitialization],
            [symbols_1.preInitialization, Enums_1.PluginHook.PreInitialization],
            [symbols_1.postInitialization, Enums_1.PluginHook.PostInitialization],
            [symbols_1.preLogin, Enums_1.PluginHook.PreLogin],
            [symbols_1.postLogin, Enums_1.PluginHook.PostLogin]
        ];
        for (const [hookSymbol, hookType] of possibleSymbolHooks) {
            const hook = Reflect.get(plugin, hookSymbol);
            if (typeof hook !== 'function')
                continue;
            this.registerHook(hook, hookType);
        }
        return this;
    }
    *values(hook) {
        for (const plugin of this.registry) {
            if (hook && plugin.type !== hook)
                continue;
            yield plugin;
        }
    }
}
exports.PluginManager = PluginManager;
