// @sapphire/framework/dist/preconditions/GuildVoiceOnly.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CorePrecondition = void 0;
const discord_js_utilities_1 = require("@sapphire/discord.js-utilities");
require("../lib/errors/Identifiers");
const Precondition_1 = require("../lib/structures/Precondition");
class CorePrecondition extends Precondition_1.AllFlowsPrecondition {
    messageRun(message) {
        return (0, discord_js_utilities_1.isVoiceChannel)(message.channel)
            ? this.ok()
            : this.error({
                identifier: "preconditionGuildVoiceOnly" /* Identifiers.PreconditionGuildVoiceOnly */,
                message: 'You can only run this message command in server voice channels.'
            });
    }
    async chatInputRun(interaction) {
        const channel = await this.fetchChannelFromInteraction(interaction);
        return (0, discord_js_utilities_1.isVoiceChannel)(channel)
            ? this.ok()
            : this.error({
                identifier: "preconditionGuildVoiceOnly" /* Identifiers.PreconditionGuildVoiceOnly */,
                message: 'You can only run this chat input command in server voice channels.'
            });
    }
    async contextMenuRun(interaction) {
        const channel = await this.fetchChannelFromInteraction(interaction);
        return (0, discord_js_utilities_1.isVoiceChannel)(channel)
            ? this.ok()
            : this.error({
                identifier: "preconditionGuildVoiceOnly" /* Identifiers.PreconditionGuildVoiceOnly */,
                message: 'You can only run this context menu command in server voice channels.'
            });
    }
}
exports.CorePrecondition = CorePrecondition;
